src/main/scala/org/allenai/pipeline/Pipeline.scala

```diff
@@ -217,143 +264,101 @@ trait Pipeline extends Logging {
+    }
 
-  protected[this] val persistedStepsByConfigKey =
-    scala.collection.mutable.Map.empty[String, Producer[_]]
+  def configured(cfg: Config) =
+    new ConfiguredPipeline {
+      val config = cfg
+    }
+}
 
-  private lazy val runOnlySteps = config.get[String]("runOnly").map(_.split(",").toSet).getOrElse(Set.empty[String])
-  private lazy val tmpOutput = config.get[String]("tmpOutput").map(s => ArtifactFactory.fromUrl(new URI(s)))
-  def isRunOnlyStep(stepName: String) = runOnlySteps(stepName)
+trait ConfiguredPipeline extends Pipeline {
+  val config: Config
+
+  override def rootOutputUrl =
+    config.get[String]("output.dir").map(s => new URI(s))
+      .getOrElse(new File(System.getProperty("user.dir")).toURI)
+
+  protected[this] def getStringList(key: String) =
+    config.get[String](key) match {
+      case Some(s) => List(s)
+      case None => config.get[Seq[String]](key) match {
+        case Some(sList) => sList
+        case None => List()
+      }
+    }
 
   override def run(rawTitle: String) = {
+    val (targets, isRunOnly) =
+      getStringList("runOnly") match {
+        case seq if seq.nonEmpty =>
+          (getStepsByName(seq), true)
+        case _ =>
+          getStringList("runUntil") match {
+            case seq if seq.nonEmpty =>
+              (getStepsByName(seq), false)
+            case _ =>
+              (persistedSteps.toList, false)
+          }
+      }
+
     config.get[Boolean]("dryRun") match {
-      case Some(true) => dryRun(new File(System.getProperty("user.dir")), rawTitle)
+      case Some(true) =>
+        val outputDir = config.get[String]("dryRunOutput")
+          .getOrElse(System.getProperty("user.dir"))
+        dryRun(new File(outputDir), rawTitle, targets)
+        List()
       case _ =>
-        config.get[String]("runOnly") match {
-          case Some(stepConfigKeys) =>
-            val (matchedNames, unmatchedNames) = runOnlySteps.partition(persistedStepsByConfigKey.contains)
-            unmatchedNames.size match {
-              case 0 =>
-                val matches = matchedNames.map(persistedStepsByConfigKey)
-                runOnly(rawTitle, matches.toList: _*)
-              case 1 =>
-                sys.error(s"Unknown step name: ${unmatchedNames.head}")
-              case _ =>
-                sys.error(s"Unknown step names: [${unmatchedNames.mkString(",")}]")
-            }
-          case _ => super.run(rawTitle)
+        if (isRunOnly) {
+          runOnly(rawTitle, targets)
+        } else {
+          runPipelineReturnResults(rawTitle, targets)
         }
     }
   }
 
-  override lazy val artifactFactory = {
-    val url = new URI(config.getString("output.dir"))
-    val path = url.getPath.stripSuffix("/")
-    val outputDirUrl = new URI(url.getScheme, url.getHost, path, null)
-    ArtifactFactory.fromUrl(outputDirUrl)
-  }
-
-  def optionallyPersist[T, A <: Artifact: ClassTag](
+  override def persist[T, A <: Artifact: ClassTag](
```

Comment:
Per email thread:

- Remove optionallyPersist() method, replace with persist()
- Behavior of persist() is to check config for an overridden output location
- If no config is found, behavior is to persist in the default location
- No longer necessary to set output.persist.XXX=true to enable persistence
- Can still disable persistence by setting output.persist.XXX=false in the config